src/server/client.rs

```diff
@@ -133,20 +133,25 @@ fn compute_job_detail(state_ref: &StateRef, job_id: JobId, include_tasks: bool)
     )
 }
 
-fn compute_job_info(state_ref: &StateRef, job_ids: Option<Vec<JobId>>) -> ToClientMessage {
+fn compute_job_info(state_ref: &StateRef, selector: JobSelector) -> ToClientMessage {
     let state = state_ref.get();
-    if let Some(job_ids) = job_ids {
-        ToClientMessage::JobInfoResponse(JobInfoResponse {
-            jobs: job_ids
-                .into_iter()
-                .filter_map(|job_id| state.get_job(job_id).map(|j| j.make_job_info()))
-                .collect(),
-        })
-    } else {
-        ToClientMessage::JobInfoResponse(JobInfoResponse {
-            jobs: state.jobs().map(|j| j.make_job_info()).collect(),
-        })
-    }
+
+    let jobs: Vec<_> = match selector {
+        JobSelector::All => state.jobs().map(|j| j.make_job_info()).collect(),
+        JobSelector::LastN(n) => {
+            let jobs = state.jobs();
+            let length = jobs.len();
+            jobs.skip(length.saturating_sub(n))
+                .map(|j| j.make_job_info())
+                .collect()
+        }
+        JobSelector::Specific(ids) => ids
+            .into_iter()
+            .filter_map(|id| state.get_job(id))
+            .map(|j| j.make_job_info())
```

Comment:
`.map(|j| j.make_job_info())` is duplicated here to avoid creating intermediate vectors or boxing the iterators.